DM-32250: minor tweaks to typing configuration

TallJimbo opened the pull request:
## Checklist

- [x] ran Jenkins
- [ ] added a release note for user-visible changes to `doc/changes`

TallJimbo committed:
Add MYPYPATH to EUPS table.

TallJimbo committed:
Update mypy config to expect typing from dependencies.

TallJimbo committed:
Fix step name in mypy actions.

TallJimbo committed:
Add py.typed to signal typing downstream.

TallJimbo commented:
@timj , do you have any guesses as to why (after my update to the `mypy.ini` here, which I think makes sense), the GitHub action mypy check can see the `utils` types, but not the `resources` ones?

timj commented:
No. There is definitely a `py.typed` file there. Maybe I need to see if `pip install` into a new local conda env does install the `py.typed` file.

TallJimbo committed:
Expand type annotation to reflect what works and is used downstream.

TallJimbo committed:
Clean up typing subtleties in storage classes and assembly.

These were revealed by running mypy on concrete assemblers in obs_base.

TallJimbo committed:
Rename mypy job inside GitHub Action.

Calling this "lint" looks like it was a copy-paste error.  Fixing it
just makes GitHub Actions outputs a bit less confusing.

timj commented on mypy.ini:
```diff
@@ -39,10 +39,25 @@ ignore_missing_imports = True
+
+[mypy-lsst.utils.*]
+ignore_missing_imports = False
+ignore_errors = True
```
Why didn't we need this before? Did the previous block mean that we weren't testing utils/resources at all before? That can't be right because when I factored out the utility code I did get mypy errors.

TallJimbo replied:
It means that when `daf_butler` imported types from `utils` or `resources` before, they were all effectively seen as `Any` by MyPy, so how much it could tell us about our usage of them was pretty limited.

timj commented on python/lsst/daf/butler/core/storageClassDelegate.py:
```diff
@@ -246,8 +246,8 @@ def getComponent(self, composite: Any, componentName: str) -> Any:
-        self, composite: Any, subset: Optional[Iterable] = None, override: bool = None
-    ) -> Dict[str, Any]:
+        self, composite: Any, subset: Optional[Iterable] = None, override: Optional[Any] = None
+    ) -> Optional[Dict[str, DatasetComponent]]:
```
I can't understand how this ever returns `None`. `components` is always a dict and is returned as a dict unless an exception is raised.

TallJimbo replied:
I'll try to remove the `Optional` and report back if I encounter a problem.  I don't recall how it got here.

TallJimbo replied:
I remembered after looking more closely at the code: the (preexisting) docstring says:
> Returns None if this is not a composite `StorageClassDelegate.storageClass`.

I assume I should remove that sentence from the docs, then?

timj replied:
I guess it changed at some point. It will seemingly always return an empty dict.

timj commented on requirements.txt:
```diff
@@ -4,7 +4,7 @@ click >7.0
 git+https://github.com/lsst/sphgeom@main#egg=lsst_sphgeom
 git+https://github.com/lsst/utils@main#egg=lsst_utils
-git+https://github.com/lsst/resources@main#egg=lsst_resources
+git+https://github.com/lsst/resources@tickets/DM-32250#egg=lsst_resources
```
Remember to back this out.

timj commented on setup.cfg:
```diff
@@ -25,7 +25,7 @@ install_requires =
   lsst_sphgeom @ git+https://github.com/lsst/sphgeom@main
   lsst_utils @ git+https://github.com/lsst/utils@main
-  lsst_resources @ git+https://github.com/lsst/resources@main
+  lsst_resources @ git+https://github.com/lsst/resources@tickets/DM-32250
```
And this.

timj reviewed (approved):
I am confused by the disassemble method returning `None` but I'm sure that's me not reading the code properly.

TallJimbo committed:
Document and annotate disassemble result as not None.
